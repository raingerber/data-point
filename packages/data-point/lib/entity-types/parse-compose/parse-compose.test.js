/* eslint-env jest */
const parseCompose = require('./parse-compose')

function expectModifier (result) {
  expect(result).toHaveProperty('type')
  expect(result).toHaveProperty('spec')
}

describe('createComposeReducer', () => {
  test('creates ModifierObject', () => {
    const result = parseCompose.createComposeReducer('map', '$a')
    expectModifier(result)
  })
})

describe('parseModifierSpec', () => {
  test('it should parse a modifier spec', () => {
    const result = parseCompose.parseModifierSpec('id', ['map'], {
      map: '$a'
    })
    expectModifier(result)
  })
})

describe('parseComposeSpec', () => {
  test('it should parse a single modifier spec', () => {
    const result = parseCompose.parseComposeSpec(
      'id',
      ['map'],
      [
        {
          map: '$a'
        }
      ]
    )

    expect(result).toHaveLength(1)
    result.forEach(expectModifier)
  })

  test('should throw error if does not contain keys', () => {
    expect(() => {
      parseCompose.parseComposeSpec('id', [], [{}])
    }).toThrowErrorMatchingSnapshot()
  })

  test('should throw error if contains more than 1 key', () => {
    expect(() => {
      parseCompose.parseComposeSpec(
        'id',
        ['map', 'filter'],
        [
          {
            map: '$a',
            filter: '$a'
          }
        ]
      )
    }).toThrowErrorMatchingSnapshot()
  })
})

describe('parse', () => {
  test('should parse a single key', () => {
    return expect(
      parseCompose.parse('id', ['map'], {
        map: '$a'
      })
    ).toMatchSnapshot()
  })

  test('should parse multiple keys inside of compose', () => {
    return expect(
      parseCompose.parse('id', ['map', 'find'], {
        compose: [
          {
            map: '$a'
          },
          {
            find: '$b'
          }
        ]
      })
    ).toMatchSnapshot()
  })

  test('should throw error for invalid key inside compose', () => {
    return expect(() =>
      parseCompose.parse('id', ['map'], {
        compose: [
          {
            find: '$a'
          }
        ]
      })
    ).toThrowErrorMatchingSnapshot()
  })

  test('should throw error when multiple keys are not inside compose', () => {
    return expect(() =>
      parseCompose.parse('id', ['map', 'find'], {
        map: '$a',
        find: '$a'
      })
    ).toThrowErrorMatchingSnapshot()
  })

  test('should throw error when the compose key and individual modifier keys are used together', () => {
    return expect(() =>
      parseCompose.parse('id', ['map', 'find'], {
        map: '$a',
        compose: [
          {
            find: '$a'
          }
        ]
      })
    ).toThrowErrorMatchingSnapshot()
  })
})
