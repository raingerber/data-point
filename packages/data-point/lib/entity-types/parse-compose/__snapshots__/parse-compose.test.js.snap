// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse should parse a single key 1`] = `
Array [
  Object {
    "spec": "$a",
    "type": "map",
  },
]
`;

exports[`parse should parse multiple keys inside of compose 1`] = `
Array [
  Object {
    "spec": "$a",
    "type": "map",
  },
  Object {
    "spec": "$b",
    "type": "find",
  },
]
`;

exports[`parse should throw error for invalid key inside compose 1`] = `"Modifier 'find' in \\"id\\" does not match any of the valid modifiers: map"`;

exports[`parse should throw error when multiple keys are not inside compose 1`] = `
"Entity \\"id\\" has invalid modifiers, when using multiple keys they should be added through the \\"compose\\" property.
Valid modifiers: map, find.
For more info: https://github.com/ViacomInc/data-point/tree/master/packages/data-point#entity-compose-reducer"
`;

exports[`parse should throw error when the compose key and individual modifier keys are used together 1`] = `
"Entity \\"id\\" is invalid; when \\"compose\\" is defined the keys: map should be inside compose.
For more info: https://github.com/ViacomInc/data-point/tree/master/packages/data-point#entity-compose-reducer"
`;

exports[`parseComposeSpec should throw error if contains more than 1 key 1`] = `"Compose Modifiers may only contain one key, but found 2 (map, filter)"`;

exports[`parseComposeSpec should throw error if does not contain keys 1`] = `"Compose Modifiers may only contain one key, but found 0"`;
